--- /tmp/libqabstractitemmodel_unpatched.hxx	2026-05-16 17:55:45.286170416 +0200
+++ /tmp/libqabstractitemmodel_fixed.hxx	2026-05-16 17:56:43.446778504 +0200
@@ -455,7 +455,9 @@
             QModelIndex* cbval3 = const_cast<QModelIndex*>(&parent_ret);
 
             QModelIndex* callback_ret = index_cb(this, cbval1, cbval2, cbval3);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return {};
     }
@@ -469,7 +471,9 @@
             QModelIndex* cbval1 = const_cast<QModelIndex*>(&child_ret);
 
             QModelIndex* callback_ret = parent_cb(this, cbval1);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return {};
     }
@@ -489,7 +493,9 @@
             QModelIndex* cbval3 = const_cast<QModelIndex*>(&idx_ret);
 
             QModelIndex* callback_ret = sibling_cb(this, cbval1, cbval2, cbval3);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractItemModel::sibling(row, column, idx);
     }
@@ -550,7 +556,9 @@
             int cbval2 = role;
 
             QVariant* callback_ret = data_cb(this, cbval1, cbval2);
-            return *callback_ret;
+            QVariant ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return {};
     }
@@ -590,7 +598,9 @@
             int cbval3 = role;
 
             QVariant* callback_ret = headerdata_cb(this, cbval1, cbval2, cbval3);
-            return *callback_ret;
+            QVariant ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractItemModel::headerData(section, orientation, role);
     }
@@ -1026,7 +1036,9 @@
             QModelIndex* cbval1 = const_cast<QModelIndex*>(&index_ret);
 
             QModelIndex* callback_ret = buddy_cb(this, cbval1);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractItemModel::buddy(index);
     }
@@ -1075,7 +1087,9 @@
             QModelIndex* cbval1 = const_cast<QModelIndex*>(&index_ret);
 
             QSize* callback_ret = span_cb(this, cbval1);
-            return *callback_ret;
+            QSize ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractItemModel::span(index);
     }
@@ -1300,7 +1314,9 @@
             int cbval2 = column;
 
             QModelIndex* callback_ret = createindex_cb(this, cbval1, cbval2);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractItemModel::createIndex(row, column);
     }
@@ -1318,7 +1334,9 @@
             uintptr_t cbval3 = static_cast<uintptr_t>(id);
 
             QModelIndex* callback_ret = createindex2_cb(this, cbval1, cbval2, cbval3);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractItemModel::createIndex(row, column, id);
     }
@@ -1722,7 +1740,9 @@
             const void* cbval3 = (const void*)data;
 
             QModelIndex* callback_ret = createindex3_cb(this, cbval1, cbval2, cbval3);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractItemModel::createIndex(row, column, data);
     }
@@ -2286,7 +2306,9 @@
             QModelIndex* cbval3 = const_cast<QModelIndex*>(&parent_ret);
 
             QModelIndex* callback_ret = index_cb(this, cbval1, cbval2, cbval3);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractTableModel::index(row, column, parent);
     }
@@ -2306,7 +2328,9 @@
             QModelIndex* cbval3 = const_cast<QModelIndex*>(&idx_ret);
 
             QModelIndex* callback_ret = sibling_cb(this, cbval1, cbval2, cbval3);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractTableModel::sibling(row, column, idx);
     }
@@ -2389,7 +2413,9 @@
             int cbval2 = role;
 
             QVariant* callback_ret = data_cb(this, cbval1, cbval2);
-            return *callback_ret;
+            QVariant ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return {};
     }
@@ -2429,7 +2455,9 @@
             int cbval3 = role;
 
             QVariant* callback_ret = headerdata_cb(this, cbval1, cbval2, cbval3);
-            return *callback_ret;
+            QVariant ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractTableModel::headerData(section, orientation, role);
     }
@@ -2825,7 +2853,9 @@
             QModelIndex* cbval1 = const_cast<QModelIndex*>(&index_ret);
 
             QModelIndex* callback_ret = buddy_cb(this, cbval1);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractTableModel::buddy(index);
     }
@@ -2874,7 +2904,9 @@
             QModelIndex* cbval1 = const_cast<QModelIndex*>(&index_ret);
 
             QSize* callback_ret = span_cb(this, cbval1);
-            return *callback_ret;
+            QSize ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractTableModel::span(index);
     }
@@ -3099,7 +3131,9 @@
             int cbval2 = column;
 
             QModelIndex* callback_ret = createindex_cb(this, cbval1, cbval2);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractTableModel::createIndex(row, column);
     }
@@ -4040,7 +4074,9 @@
             QModelIndex* cbval3 = const_cast<QModelIndex*>(&parent_ret);
 
             QModelIndex* callback_ret = index_cb(this, cbval1, cbval2, cbval3);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractListModel::index(row, column, parent);
     }
@@ -4060,7 +4096,9 @@
             QModelIndex* cbval3 = const_cast<QModelIndex*>(&idx_ret);
 
             QModelIndex* callback_ret = sibling_cb(this, cbval1, cbval2, cbval3);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractListModel::sibling(row, column, idx);
     }
@@ -4129,7 +4167,9 @@
             int cbval2 = role;
 
             QVariant* callback_ret = data_cb(this, cbval1, cbval2);
-            return *callback_ret;
+            QVariant ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return {};
     }
@@ -4169,7 +4209,9 @@
             int cbval3 = role;
 
             QVariant* callback_ret = headerdata_cb(this, cbval1, cbval2, cbval3);
-            return *callback_ret;
+            QVariant ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractListModel::headerData(section, orientation, role);
     }
@@ -4565,7 +4607,9 @@
             QModelIndex* cbval1 = const_cast<QModelIndex*>(&index_ret);
 
             QModelIndex* callback_ret = buddy_cb(this, cbval1);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractListModel::buddy(index);
     }
@@ -4614,7 +4658,9 @@
             QModelIndex* cbval1 = const_cast<QModelIndex*>(&index_ret);
 
             QSize* callback_ret = span_cb(this, cbval1);
-            return *callback_ret;
+            QSize ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractListModel::span(index);
     }
@@ -4839,7 +4885,9 @@
             int cbval2 = column;
 
             QModelIndex* callback_ret = createindex_cb(this, cbval1, cbval2);
-            return *callback_ret;
+            QModelIndex ret = *callback_ret;
+            delete callback_ret;
+            return ret;
         }
         return QAbstractListModel::createIndex(row, column);
     }
